frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,845 @@
+
+    def set_qubit_to_wire(self, idx, qubit):
+        """Keep track of which wire this qubit variable is."""
+        self.qubit_to_wire_map[qubit] = idx
+
+    def get_wire_for_qubit(self, qubit):
+        """Get the wire for this qubit."""
+        return self.qubit_to_wire_map[qubit]
+
+    def read(self, var):
+        """Read the value of variable var."""
+        if isinstance(var, jax.core.Literal):
+            return var.val
+        if self.variable_map.get(var):
+            var = self.variable_map[var]
+        return self.env[var]
+
+    def write(self, var, val):
+        """var = val."""
+        if self.variable_map.get(var):
+            var = self.variable_map[var]
+        self.env[var] = val
+
+    def replace(self, original, new):
+        """Replace original variable with new variable."""
+        self.variable_map[original] = new
+
+    def get_new_count(self):
+        """Increase count and return."""
+        self.count += 1
+        return self.count
+
+    def new_variable(self):
+        """Convenience to get a new variable of a given type.
+
+        These are not JAX variables. This can be any hashable.
+        We are using integers as a small level of indirection.
+        Each new variable is a new integer.
+
+        This is pretty much what JAX does since each variable gets
+        assigned a count and a type. But it is also what anyone else does.
+        (Like LLVM/MLIR %0, %1, ...)
+
+        TODO(@erick-xanadu): I think the type information
+        would also be good to have here, but it is unnecessary at the
+        moment.
+        """
+        return self.get_new_count()
+
+
+def change_device_to_cuda_device(ctx):
+    """Map Catalyst's qdevice_p primitive to its equivalent CUDA-quantum primitive
+    as defined in this file.
+
+    From here we get shots as well.
+    """
+
+    # The device here might also have some important information for
+    # us. For example, the number of shots.
+
+    qdevice_eqn = get_instruction(ctx.jaxpr, qdevice_p)
+
+    # These parameters are stored in a json-like string.
+    # We first convert the string to json.
+    json_like_string = qdevice_eqn.params["rtd_kwargs"]
+    json_like_string = json_like_string.replace("'", '"')
+    json_like_string = json_like_string.replace("True", "true")
+    json_string = json_like_string.replace("False", "false")
+
+    # Finally, we load it
+    parameters = json.loads(json_string)
+
+    # Now we have the number of shots.
+    # Shots are specified in PL at the very beginning, but in cuda
+    # shots are not needed until the very end.
+    # So, we will just return this variable
+    # and it is the responsibility of the caller to propagate this information.
+    shots = parameters.get("shots")
+
+    device_name = parameters.get("rtd_name", "softwareq.qpp")
+
+    # TODO(@erick-xanadu) as more devices become available
+    # map the names here.
+    target_map = {"softwareq.qpp": "qpp-cpu"}
```

Comment:
This is not a problem right now without the other devices (is any additional effort to add them or are they just plug and play?), but the map will not be static since it depends on device options (like `mps` and `multi_gpu`).

Reply:
Can we add the target map to the device, somehow? E.g.,

```python
class SoftwareQQPP(CudaBaseDevice):
    short_name = "softwareq.qpp"
    nv_target = "qpp-cpu"
```

Assuming that `dev.nv_target` ends up in the JAXPR, and can be caught/transformed.

The reason I like this is it keeps information about the device in the device; the interpreter would then simply be in charge of making sure it catches this information.

Reply:
It may? also solve @dime10's issue, because you could do this:

```python
class CuStateVec(CudaBaseDevice):

    short_name = "nvidia.custatevec"

    @property
    def nv_target(self):
        if self.multi_gpu:
            return "nvidia-mgpu"
        return "nvidia"
```

Reply:
I like this idea, but can it wait until we can test the execution on a machine? I don't want to enable something just to find out that it doesn't work for some other reason (like conflicts of symbols).